other/shared_company/src/proposal.rs

```diff
@@ -0,0 +1,117 @@
+use scrypto::prelude::*;
+// A proposal that unlocks funds if accepted.
+blueprint! {
+struct Proposal{
+            // Fund which are looked for the vote
+            cost_vault: Vault,
+            // If proposal is accepted: Funds are send to this adress
+            destination_adress_funds: Address,
+            // What are users voting on?
+            reason: String,
+            // The token that is needed to vote.
+            company_voting_token: Vault,
+            // Material for voting. If the vault could count, yes and no counter wouldnt be needed
+            replacement_tokens_type_yes: Vault,
+            replacement_tokens_type_no: Vault,
+            yes_counter: Decimal,
+            no_counter: Decimal,
+            // The votes needed for the vote to succeed
+            needed_votes: Decimal,
+            // When this epoch is reached and try_solve() is called,
+            //no more voting will be possible and cost_vault will be send to owners address
+            end_epoch: u64,
+            fund_owner_adress: Address,
+    }
+
+            impl Proposal{
+
+                /// creates a new instance of a proposal
+                pub fn new(cost: Bucket, destination_adress_funds: Address, reason: String, admin_adress: Address, end_epoch: u64,
+                    needed_votes: Decimal, company_voting_token_resource_def: ResourceDef)-> Component {
+
+                    // The token that the user gets in exchange for their "yes" voting.
+                    let replacement_token_yes_resource_def = ResourceBuilder::new()
+                       .metadata("name", "Replacement token yes").metadata("symbol", "RTY")
+                    .new_token_fixed(1_000_000);
+
+                     // The token that the user gets in exchange for their "no" voting.
+                     let replacement_token_no_resource_def = ResourceBuilder::new()
+                     .metadata("name", "Replacement token no").metadata("symbol", "RTN")
+                  .new_token_fixed(1_000_000);
+
+                    // the vault that holds the costs that are associated with the proposal
+                    let cost_vault = Vault::new(cost.resource_def());
+                    // fills the cost vault
+                    cost_vault.put(cost);
+
+                  Self {
+                    cost_vault: cost_vault,
+                    destination_adress_funds: destination_adress_funds, reason: reason,
+                    company_voting_token : Vault::new(company_voting_token_resource_def),
+                    replacement_tokens_type_yes: Vault::with_bucket(replacement_token_yes_resource_def),
+                    replacement_tokens_type_no: Vault::with_bucket(replacement_token_no_resource_def),
+                    yes_counter: Decimal(0.0 as i128),
```

Comment:
Instead of creating two tokens (Yes and No), why not create a single Vote badge. When the user shows their vote badge they say wether they vote yes or no. Then the component burns the badge so that they can only vote once